decimal/bcd.go

```diff
@@ -0,0 +1,210 @@
+//
+//     |  4 bits           |  4 bits           |
+//        = the last digit    = nibble
+//
+// The final nibble represents the number's sign: 0x0a, 0x0c, 0x0e, 0x0f stand
+// for plus, 0x0b and 0x0d stand for minus.
+//
+// Examples:
+//
+// The decimal -12.34 will be encoded as 0xd6, 0x01, 0x02, 0x01, 0x23, 0x4d:
+//
+// |MP_EXT (fixext 4) | MP_DECIMAL | scale |  1   |  2,3 |  4 (minus) |
+// |       0xd6       |    0x01    | 0x02  | 0x01 | 0x23 | 0x4d       |
+//
+// The decimal 0.000000000000000000000000000000000010 will be encoded as
+// 0xc7, 0x03, 0x01, 0x24, 0x01, 0x0c:
+//
+// | MP_EXT (ext 8) | length | MP_DECIMAL | scale |  1   | 0 (plus) |
+// |      0xc7      |  0x03  |    0x01    | 0x24  | 0x01 | 0x0c     |
+//
+// See also:
+//
+// * MessagePack extensions https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/
+//
+// * An implementation in C language https://github.com/tarantool/decNumber/blob/master/decPacked.c
+package decimal
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+var mapRuneToByteSign = map[rune]byte{
+	'+': 0x0c,
+	'-': 0x0d,
+}
+
+var isNegative = map[byte]bool{
+	0x0a: false,
+	0x0b: true,
+	0x0c: false,
+	0x0d: true,
+	0x0e: false,
+	0x0f: false,
+}
+
+// EncodeStringToBCD converts a string buffer to BCD Packed Decimal.
+//
+// The number is converted to a BCD packed decimal byte array, right aligned in
+// the BCD array, whose length is indicated by the second parameter. The final
+// 4-bit nibble in the array will be a sign nibble, 0x0a for "+" and 0x0b for
+// "-". Unused bytes and nibbles to the left of the number are set to 0. scale
+// is set to the scale of the number (this is the exponent, negated).
+func EncodeStringToBCD(buf string) ([]byte, error) {
+	sign := '+' // By default number is positive.
+	if res := strings.HasPrefix(buf, "-"); res {
+		sign = '-'
+	}
+
+	// Calculate a number of digits in the decimal number. Leading and
+	// trailing zeros do not count.
+	// Examples:
+	// 0.0000000000000001 - 1 digit
+	// 00012.34           - 4 digits
+	// 0.340              - 2 digits
+	s := strings.ReplaceAll(buf, "-", "") // Remove a sign.
+	s = strings.ReplaceAll(s, "+", "")    // Remove a sign.
+	s = strings.ReplaceAll(s, ".", "")    // Remove a dot.
+	s = strings.TrimLeft(s, "0")          // Remove leading zeros.
+	c := len(s)
+
+	// Fix a case with a single 0.
+	if c == 0 {
+		s = "0" //nolint
+		c = 1
+	}
+
+	// The first nibble should contain 0, if the decimal number has an even
+	// number of digits. Therefore highNibble is false when decimal number
+	// is even.
+	highNibble := true
+	if c%2 == 0 {
+		highNibble = false
+	}
+	scale := 0 // By default decimal number is integer.
+	var byteBuf []byte
+	for i, ch := range buf {
+		// Skip leading zeros.
+		if (len(byteBuf) == 0) && ch == '0' {
+			continue
+		}
+		if (i == 0) && (ch == '-' || ch == '+') {
+			continue
+		}
+		// Calculate a number of digits after the decimal point.
+		if ch == '.' {
```

Comment:
`&& scale == 0` ?

Reply:
A number with more than a single point is not a correct number. However, I got your idea.

```diff
diff --git a/decimal/bcd.go b/decimal/bcd.go
index bf8e0d3..92a5f89 100644
--- a/decimal/bcd.go
+++ b/decimal/bcd.go
@@ -110,6 +110,9 @@ func EncodeStringToBCD(buf string) ([]byte, error) {
 		if (i == 0) && (ch == '-' || ch == '+') {
 			continue
 		}
+		if ch == '.' && scale != 0 {
+			return nil, fmt.Errorf("Number contains more than one point")
+		}
 		// Calculate a number of digits after the decimal point.
 		if ch == '.' {
 			scale = len(buf) - i - 1
diff --git a/decimal/decimal_test.go b/decimal/decimal_test.go
index f690ced..b8890c1 100644
--- a/decimal/decimal_test.go
+++ b/decimal/decimal_test.go
@@ -145,6 +145,21 @@ func TestMPEncodeDecode(t *testing.T) {
 	}
 }
 
+func TestEncodeStringToBCDIncorrectNumber(t *testing.T) {
+	referenceErrMsg := "Number contains more than one point"
+	var numString = "0.1.0"
+	buf, err := EncodeStringToBCD(numString)
+	if err == nil {
+		t.Fatalf("It is possible to encode a string with incorrect number")
+	}
+	if buf != nil {
+		t.Fatalf("buf is not nil on attempt to encode a string with incorrect number")
+	}
+	if err.Error() != referenceErrMsg {
+		t.Fatalf("Incorrect error message on attempt to encode a string with incorrect number")
+	}
+}
+
 func TestEncodeMaxNumber(t *testing.T) {
 	referenceErrMsg := "msgpack: decimal number is bigger than maximum supported number (10^38 - 1)"
 	decNum := decimal.New(1, 38) // 10^38
```